File: flutter_app/lib/models/health_log.dart
/// Types of health logs that can be recorded.
enum HealthLogType { mood, symptoms, meals, water, exercise, sleep, general }

/// Display label for type (formatting only; add new enum values without relisting here).
String formatHealthLogTypeDisplay(HealthLogType type) =>
    type.name[0].toUpperCase() + type.name.substring(1);

/// Health log model for quick logs and daily entries.
class HealthLog {
  final String id;
  final HealthLogType type;
  final String description;
  final String? note;
  final DateTime createdAt;
  final String? emoji;
  final double? waterTotal;
  final double? waterDelta;
  final double? waterGoal;
  final double? sleepHours;

  HealthLog({
    required this.id,
    required this.type,
    required this.description,
    required this.createdAt,
    this.note,
    this.emoji,
    this.waterTotal,
    this.waterDelta,
    this.waterGoal,
    this.sleepHours,
  });

  HealthLog copyWith({
    String? id,
    HealthLogType? type,
    String? description,
    String? note,
    DateTime? createdAt,
    String? emoji,
    double? waterTotal,
    double? waterDelta,
    double? waterGoal,
    double? sleepHours,
  }) {
    return HealthLog(
      id: id ?? this.id,
      type: type ?? this.type,
      description: description ?? this.description,
      createdAt: createdAt ?? this.createdAt,
      note: note ?? this.note,
      emoji: emoji ?? this.emoji,
      waterTotal: waterTotal ?? this.waterTotal,
      waterDelta: waterDelta ?? this.waterDelta,
      waterGoal: waterGoal ?? this.waterGoal,
      sleepHours: sleepHours ?? this.sleepHours,
    );
  }

  /// Date-only (midnight) for filtering by calendar day.
  DateTime get dateOnly =>
      DateTime(createdAt.year, createdAt.month, createdAt.day);

  /// Whether this log has water progress information.
  bool get hasProgress => waterTotal != null && waterGoal != null;

  /// Clamp ratio between 0 and 1 for progress display.
  double get progressRatio {
    if (!hasProgress || waterGoal == 0) return 0;
    return (waterTotal! / waterGoal!).clamp(0, 1);
  }
}

File: react_native_app/models/HealthLog.ts
// =============================================================================
// HEALTH LOG MODEL
// =============================================================================
// Domain model for health-related entries (mood, water, meals, blood pressure,
// etc.). Used by HealthLogProvider and Health Logs / Health Timeline screens.
// =============================================================================

export enum HealthLogType {
  mood = 'mood',
  symptoms = 'symptoms',
  meals = 'meals',
  water = 'water',
  exercise = 'exercise',
  sleep = 'sleep',
  general = 'general',
  bloodPressure = 'bloodPressure',
  heartRate = 'heartRate',
}

export function formatHealthLogTypeDisplay(type: HealthLogType): string {
  switch (type) {
    case HealthLogType.bloodPressure:
      return 'Blood Pressure';
    case HealthLogType.heartRate:
      return 'Heart Rate';
    default:
      const name = type;
      return name.charAt(0).toUpperCase() + name.slice(1);
  }
}

/** Blood pressure category per AHA guidelines (systolic/diastolic mmHg). */
export function bloodPressureCategoryLabel(systolic: number, diastolic: number): string {
  if (systolic >= 180 || diastolic >= 120) return 'Hypertensive crisis';
  if (systolic >= 140 || diastolic >= 90) return 'High (Stage 2)';
  if (systolic >= 130 || diastolic >= 80) return 'High (Stage 1)';
  if (systolic >= 120 && diastolic < 80) return 'Elevated';
  return 'Normal';
}

/** Heart rate category for resting adult (bpm). */
export function heartRateCategoryLabel(bpm: number): string {
  if (bpm < 60) return 'Bradycardia';
  if (bpm <= 100) return 'Normal';
  return 'Tachycardia';
}

export interface HealthLog {
  id: string;
  type: HealthLogType;
  description: string;
  note?: string | null;
  createdAt: Date;
  emoji?: string | null;
  waterTotal?: number | null;
  waterDelta?: number | null;
  waterGoal?: number | null;
  sleepHours?: number | null;
  systolic?: number | null;
  diastolic?: number | null;
  heartRateBpm?: number | null;
}

/** Date-only (midnight) for filtering by calendar day. */
export function getHealthLogDateOnly(log: HealthLog): Date {
  const d = log.createdAt;
  return new Date(d.getFullYear(), d.getMonth(), d.getDate());
}

export function healthLogHasProgress(log: HealthLog): boolean {
  return log.waterTotal != null && log.waterGoal != null && log.waterGoal > 0;
}

export function healthLogProgressRatio(log: HealthLog): number {
  if (!healthLogHasProgress(log) || !log.waterGoal) return 0;
  return Math.min(1, Math.max(0, (log.waterTotal ?? 0) / log.waterGoal));
}
